File: HW2/Simulator.py
_instructions = []
_asm_name = ['noop', 'lw', 'lb', 'sw', 'mv', 'add', 'addf',
             'or', 'and', 'xor', 'srl', 'beq', 'halt']


class Float:
    def __init__(self, num):
        self.sign = num >> 7
        self.exp = (num & (0b1110000)) >> 4
        self.man = num & 0b1111

    def __scale_up_exp(self, b):
        if self.exp > b.exp:
            b.man >>= self.exp - b.exp
            b.exp = self.exp
        else:
            self.man >>= b.exp - self.exp
            self.exp = b.exp

    def __add__(self, b):
        self.__scale_up_exp(b)
        if self.sign == b.sign:
            sum = self.man + b.man
            exp = self.exp
            if sum > 0b1111 and exp < 0b111:
                sum >>= 1 # truncate
                exp += 1
            return self.sign << 7 | exp << 4 | sum & 0b1111 # exp == 0b111, make it overflow (normalized)
        else:
            if self.man == b.man:
                return self.exp << 4 # 0 with exponent
            else:
                return (self.sign if self.man > b.man else b.sign) << 7 | self.exp << 4 | abs(self.man - b.man)

class Simulator:
    def __init__(self):
        print()
        if not getattr(self, 'instructions', None):
            self.instructions = _instructions
        self.memory = None
        self.registers = [0 for i in range(16)]
        self.pc = 0

    def loadMemory(self, path):
        print("loadMemory")
        with open(path, 'rb') as file:
            self.memory = bytearray(file.read())

    def loadAssembly(self, path):
        print("loadAssembly")
        if not getattr(self, 'asm', None):
            self.asm = {name: index for index, name in enumerate(_asm_name)}
        self.memory = bytearray(256)
        with open(path, 'r') as file:
            for index, line in enumerate(file):
                code = line.strip().lower().split(' ')
                inst = code[0]
                self.memory[2*index] = self.asm[inst] << 4
                if inst == 'mv':
                    self.memory[2*index + 1] = int(''.join(code[1:]), 16)
                elif inst == 'srl':
                    self.memory[2*index] |= int(code[1], 16)
                    self.memory[2*index + 1] = int(code[-1], 16)
                elif inst != 'halt':  # if is halt, C000, inst is already set
                    code = ''.join(code[1:])  # skip first (instruction)
                    self.memory[2*index] |= int(code[0], 16)
                    self.memory[2*index + 1] = int(code[1:], 16)

    def storeMemory(self, path):
        assert self.memory, "Memory is not loaded, call Simulator.loadMemory first!"
        print("storeMemory")
        with open(path, 'wb') as file:
            file.write(self.memory)

    def simulate(self):
        assert self.memory, "Memory is not loaded, call Simulator.loadMemory first!"
        self.registers = [0 for i in range(16)]
        print("simulate")
        self.pc = 0
        while not self.instructions[((self.memory[self.pc] & 0xF0) >> 4)](self, (self.memory[self.pc] & 0xF) << 8 | self.memory[self.pc+1]):
            self.pc += 2

    # Helper functions
    def instruction(*masks):  # Concatenate 2 bytes, use argument as mask
        masks = [(mask, 4*'{:x}'.format(mask).count('0')) for mask in masks]
        def wrapper(func):
            global _instructions

            def ret(self, operand):
                return func(self, *((operand & mask) >> shift for mask, shift in masks))
            _instructions.append(ret)
            return ret
        return wrapper

    @instruction()
    def noop(self):
        return False
    
    @instruction(0xF00, 0xFF)
    # Extract 2nd bit, 3rd to 4th bit from max 0xFFFF int
    def laod_addr(self, reg, addr):
        self.registers[reg] = self.memory[addr]

    @instruction(0xF00, 0xFF)
    def load_const(self, reg, value):
        self.registers[reg] = value

    @instruction(0xF00, 0xFF)
    def store(self, reg, addr):
        self.memory[addr] = self.registers[reg]

    @instruction(0xF0, 0xF)
    def move(self, src, dest):
        self.registers[dest] = self.registers[src]

    @instruction(0xF00, 0xF0, 0xF)
    # Extract 2nd bit, 3rd and 4th bit from max 0xFFFF int
    def add_signed(self, dest, a, b):
        self.registers[dest] = (self.registers[a] + self.registers[b]) & 0xFF

    @instruction(0xF00, 0xF0, 0xF)
    def add_float(self, dest, a, b):
        self.registers[dest] = Float(self.registers[a]) + Float(self.registers[b])

    @instruction(0xF00, 0xF0, 0xF)
    def binary_or(self, dest, a, b):
        self.registers[dest] = self.registers[a] | self.registers[b]

    @instruction(0xF00, 0xF0, 0xF)
    def binary_and(self, dest, a, b):
        self.registers[dest] = self.registers[a] & self.registers[b]

    @instruction(0xF00, 0xF0, 0xF)
    def binary_xor(self, dest, a, b):
        self.registers[dest] = self.registers[a] ^ self.registers[b]

    @instruction(0xF00, 0xF)
    def rotate_r(self, src, times):
        times = times % 8  # rotate a 8-bit int over 8 is useless
        self.registers[src] = self.registers[src] >> times | (
            self.registers[src] & ((1 << times) - 1)) << (8 - times)

    @instruction(0xF00, 0xFF)
    def jeq(self, reg, value):
        if self.registers[reg] == self.registers[0]:
            self.pc = value - 2 # pc += 2 when instruction is executed

    @instruction()
    def halt(self):
        return True  # Other instructions returns None

File: HW2/Simulator.cpp
#include "Simulator.h"
#include <algorithm>
#include <iostream>
#include <fstream>
#include <cstring>
#include <cctype>
#include <cassert>
#define _2_34(num) unsigned reg = (num>>8) & 0xF, target = num & 0xFF
#define _2_3_4(num) unsigned dest = (num>>8) & 0xF, a = (num>>4) & 0xF, b = num & 0xF
#define _3_4(num) unsigned src = (num>>4) & 0xF, dest = num & 0xF
#define _2_4(num) unsigned reg = (num>>8) & 0xF, times = num & 0xF

bool Simulator::loadMemory(const std::string& path) {
    std::ifstream ifs(path, std::ios::in | std::ios::binary);
    return ifs && ifs.read(reinterpret_cast<char *>(_memory), sizeof(_memory));
}

bool Simulator::loadAssembly(const std::string& path) {
    std::ifstream ifs(path, std::ios::in | std::ios::binary);
    if(!ifs)
        return false;
    std::memset(_memory, 0, sizeof(_memory));
    char buffer[5]; // longest word is 4 char
    for(unsigned char pos = 0; ifs >> buffer; pos += 2) {
        for(char* temp=buffer; *temp; ++temp)
            *temp = std::tolower(*temp);
        unsigned index = std::find_if(_instructionLiteral, _instructionLiteral+13, [&](const char* i){ return std::strcmp(i, buffer) == 0;}) - _instructionLiteral;
        assert(index < 13); // Invalid instruction
        _memory[pos] = index << 4; // Store op code
        if(index == 0xc) // halt
            break;
        int a, b, c;
        ifs >> std::hex >> a >> b;
        switch(index) {
            case 4: // mv
                _memory[pos + 1] = a << 4 | b;
                break;
            case 0xa: // srl;
            case 0xb: // beq
                ifs >> c;
                _memory[pos] |= b;
                _memory[pos + 1] = c;
                break;
            case 1:
            case 2:
            case 3:
                _memory[pos] |= a;
                _memory[pos + 1] = b;
                break;
            default:
                ifs >> c;
                _memory[pos] |= a;
                _memory[pos + 1] = b << 4 | c;
        }
    }
    return true;
}

bool Simulator::storeMemory(const std::string& path) {
    std::ofstream ofs(path, std::ios::out | std::ios::binary);
    return ofs && ofs.write(reinterpret_cast<char *>(_memory), sizeof(_memory));
}

bool Simulator::simulate() {
    std::memset(_registers, 0, sizeof(_registers));
    for(_pc = 0; !(this->*_instructions[(_memory[_pc] & 0xF0) >> 4])((_memory[_pc] << 8) | _memory[_pc+1]); _pc += 2);
    return true;
}

bool Simulator::_noop(uint16_t operand) {
    return false;
}

bool Simulator::_lw(uint16_t operand) {
    _2_34(operand);
    _registers[reg] = _memory[target];
    return false;
}

bool Simulator::_lb(uint16_t operand) {
    _2_34(operand);
    _registers[reg] = target;
    return false;
}

bool Simulator::_sw(uint16_t operand) {
    _2_34(operand);
    _memory[target] = _registers[reg];
    return false;
}

bool Simulator::_mv(uint16_t operand) {
    _3_4(operand);
    _registers[dest] = _registers[src];
    return false;
}

bool Simulator::_add(uint16_t operand) {
    _2_3_4(operand);
    _registers[dest] = _registers[a] + _registers[b];
    return false;
}

bool Simulator::_addf(uint16_t operand) {
    _2_3_4(operand);
    _registers[dest] = static_cast<unsigned char>(Float(_registers[a]) + Float(_registers[b]));
    return false;
}

bool Simulator::_or(uint16_t operand) {
    _2_3_4(operand);
    _registers[dest] = _registers[a] | _registers[b];
    return false;
}

bool Simulator::_and(uint16_t operand) {
    _2_3_4(operand);
    _registers[dest] = _registers[a] & _registers[b];
    return false;
}

bool Simulator::_xor(uint16_t operand) {
    _2_3_4(operand);
    _registers[dest] = _registers[a] ^ _registers[b];
    return false;
}

bool Simulator::_srl(uint16_t operand) {
    _2_4(operand);
    times %= 8;
    _registers[reg] = _registers[reg] >> times | (_registers[reg] & ((1 << times) - 1)) << (8 - times);
    return false;
}

bool Simulator::_beq(uint16_t operand) {
    _2_34(operand);
    if(_registers[reg] == _registers[0])
        _pc = target - 2;
    return false;
}

bool Simulator::_halt(uint16_t operand) {
    return true;
}

Simulator::Instruction Simulator::_instructions[] = {
    &Simulator::_noop, &Simulator::_lw, &Simulator::_lb, &Simulator::_sw,
    &Simulator::_mv, &Simulator::_add, &Simulator::_addf, &Simulator::_or,
    &Simulator::_and, &Simulator::_xor, &Simulator::_srl, &Simulator::_beq,
    &Simulator::_halt
};

char Simulator::_instructionLiteral[][5] = {
    "noop", "lw", "lb", "sw", "mv", "add", "addf", "or", "and", "xor", "srl", "beq", "halt"
};